hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/invalidation/InvalidationMetaDataFetcher.java

```diff
@@ -27,80 +29,116 @@
+        for (Map.Entry<Member, InternalCompletableFuture> entry : futureByMember.entrySet()) {
+            Member member = entry.getKey();
+            InternalCompletableFuture future = entry.getValue();
+
+            extractMemberMetadata(member, future, resultHolder);
 
             initUuid(resultHolder.partitionUuidList, handler);
             initSequence(resultHolder.namePartitionSequenceList, handler);
         }
     }
 
-    public final void process(InternalCompletableFuture future, ConcurrentMap<String, RepairingHandler> handlers) {
-        try {
-            ResultHolder resultHolder = new ResultHolder();
-            extractAndPopulateResult(future, resultHolder);
-            repairUuids(resultHolder.partitionUuidList, handlers);
-            repairSequences(resultHolder.namePartitionSequenceList, handlers);
-        } catch (Exception e) {
-            if (logger.isWarningEnabled()) {
-                logger.warning("Can't fetch invalidation meta-data [" + e.getClass().getSimpleName() + "] " + e.getMessage());
-            }
-        }
-    }
-
     public final void fetchMetadata(ConcurrentMap<String, RepairingHandler> handlers) {
         if (handlers.isEmpty()) {
             return;
         }
 
-        List<String> mapNames = getNames(handlers);
-        List<InternalCompletableFuture> futures = scanMembers(mapNames);
-        for (InternalCompletableFuture future : futures) {
-            process(future, handlers);
+        List<String> dataStructureNames = getDataStructureNames(handlers);
+        Map<Member, InternalCompletableFuture> futureByMember = fetchMembersMetadataFor(dataStructureNames);
+        for (Map.Entry<Member, InternalCompletableFuture> entry : futureByMember.entrySet()) {
+            Member member = entry.getKey();
+            InternalCompletableFuture future = entry.getValue();
+            processMemberMetadata(member, future, handlers);
+        }
+    }
+
+    protected abstract Collection<Member> getDataMembers();
+
+    protected abstract void extractMemberMetadata(Member member,
+                                                  InternalCompletableFuture future,
+                                                  MetadataHolder metadataHolder) throws Exception;
+
+    protected abstract InternalCompletableFuture fetchMetadataOf(Address address, List<String> names);
+
+    private Map<Member, InternalCompletableFuture> fetchMembersMetadataFor(List<String> names) {
+        Collection<Member> members = getDataMembers();
+        if (members.isEmpty()) {
+            return Collections.emptyMap();
+        }
+
+        Map<Member, InternalCompletableFuture> futureByMember = createHashMap(members.size());
+        for (Member member : members) {
+            Address address = member.getAddress();
+            try {
+                futureByMember.put(member, fetchMetadataOf(address, names));
+            } catch (Exception e) {
+                if (logger.isWarningEnabled()) {
```

Comment:
Maybe we can eliminate some logging noise by logging `HazelcastInstanceNotActiveException` at `FINEST` level? If a member is gone, then there's not much we can do about it.

Also, would it make sense to keep the previous abstraction (`scanMembers`) so at least the member-side implementation could take advantage of `InvocationUtil#invokeOnStableClusterSerial` to get a clean run on all members even when cluster topology changes?

Reply:
- Added `HazelcastInstanceNotActiveException` as you suggested
- Having `InvocationUtil#invokeOnStableClusterSerial` is not crucial for this feature. This is a periodically running task, subsequent runs can fix any unexpected situation.